### ISSUE TITLE
vs2017 action results in solution that targets Windows SDK 8.1

### ISSUE THREAD

**Issue by JohannesMP** [premake/premake-core#935]
On windows 10 with a c++ project, using `premake5 vs2017` will result in a solution that targets the Windows SDK 8.1, not the new 10.0.* SDK Versions.

Example error message when trying to build a bog standard 'Hello World' C++ solution created with Premake5 (compiled from source on commit 66362c0 ):

`1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Platforms\Win32\PlatformToolsets\v141\Toolset.targets(34,5): error MSB8036: The Windows SDK version 8.1 was not found. Install the required version of Windows SDK or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solution".`

This will always happen on a machine that only has VS2017 installed, without the 8.1 SDK, which is the default state. On such a machine, the 'Retarget Projects' right click option shows the following:
![](https://i.imgur.com/Yflukv7.png).

While this is a quickfix, I feel that the default behavior for the `vs2017` action should be to use the 10.0 SDK that is guaranteed to be installed, not the 8.1 SDK that will become less and less used as more devs upgrade to using only 2017.

We should not assume that someone running the `vs2017` action has VS2015 or the older SDK installed.

------------

Edit: I just saw #450 and the addition of the [`systemversion`](https://github.com/premake/premake-core/wiki/systemversion), so at least I have a nicer workaround now with `systemversion("10.0.15063.0")`, but I still feel that it would be a more reasonable default than 8.1.

**Comment by dcourtois** [comment]
I have the same problem. And even worse in my opinion is that bootstraping Premake with vs2017 doesn't work : dev /upgrade will not update the SDK version which will still point to 8.1 :/

Basically doing this will not work:

```
"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvarsall.bat"
nmake -f Bootstrap.mak windows
```

**Comment by JohannesMP** [comment]
If it helps, I ended up adding this to my premake5.lua file: 

````lua
function os.winSdkVersion()
    local reg_arch = iif( os.is64bit(), "\\Wow6432Node\\", "\\" )
    local sdk_version = os.getWindowsRegistry( "HKLM:SOFTWARE" .. reg_arch .."Microsoft\\Microsoft SDKs\\Windows\\v10.0\\ProductVersion" )
    if sdk_version ~= nil then return sdk_version end
end
````
The code for which I believe I found in another issue or commend on a commit (can't find it again right now, but credit to them)

Then in the project configuration you can just do: 
````lua
filter {"system:windows", "action:vs*"}
    systemversion(os.winSdkVersion() .. ".0")
````

This has worked for me on different windows 8.1 and 10 machines what each have different SDK's installed

**Comment by dcourtois** [comment]
Thanks, works like a charm on my projects ! Maybe this should be the default for the `vs2017` action ? Visual Studio 2017 doesn't seem to support any SDK < 10.0, so maybe the vs2017 action should initialize the `systemversion` from the registry like your snippet does ? (if any of the authors can share their view on this ?)

**Comment by samsinsane** [comment]
The below is just my opinion, and the others may disagree with me.

> Visual Studio 2017 doesn't seem to support any SDK < 10.0

The installer provides the option to install the 8.1 SDK. So it does support it, just not as part of the "default" installation, but it is there, and others might choose to install it.

> so maybe the vs2017 action should initialize the `systemversion` from the registry like your snippet does ?

This won't work when generating the project files on another machine, this is something that @starkos feels very strongly about. Additionally, IIRC this registry key is actually the last SDK installed, not the latest SDK available. There's only one version, yet I have two versions installed on my desktop and three installed on my tablet - where do you get the oldest SDK installed?

 - Is there a version of the 10.0 SDK that is guaranteed to be installed? I don't think so, since my desktop has `10.0.15063.0` and `10.0.16299.0` but my tablet only has `10.0.10150.0`, `10.0.10240.0` and `10.0.10586.0`.
 - Will a future update to VS2017 change the default SDK version and will people open issues because Premake is doing the wrong thing again?
 - What about VS2015? It also supports specifying the SDK version. It would be inconsistent to only have VS2017 pick a random SDK.
 - Defaulting to the latest version installed can cause issues if functions are added/removed between the versions. (I have no idea if there's any API differences)
 - This particular API doesn't support defaulting to what the tool uses. We would need to add a way to do this in the event you don't want the element to be emit. For example, when you're using the versions of VS2015 prior to Update 3.
 - What about those users that opt to not install the Windows 10 SDK with VS2017? The registry key is only for the Windows 10 SDK, so likely wouldn't exist.

Having said all of this, I find it quite infuriating that I can't bootstrap Premake with VS2017 - nor can I build most of my projects without modifying the build scripts after reinstalling my OS and only installing VS2017 with the Win10 SDK. All because Microsoft decided to have VS2017 default to the 8.1 SDK even when it's not installed. 😡

**Comment by dcourtois** [comment]
> The installer provides the option to install the 8.1 SDK. So it does support it, just not as part of the "default" installation, but it is there, and others might choose to install it.

Yes, you can install 8.1, but as far as I've tested, if you actually try to build a simple app using Visual Studio 2017 with the v140 platform toolset (the default) and 8.1 Windows SDK, it won't compile with the following errors:

```
c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.11.25503\include\cstddef(5): fatal error C1083: Cannot open include file: 'stddef.h': No such file or directory
```

Basically, platform toolset 140 is not compatible with Windows SDK 8.1 (and probably anything lower) so it doesn't make any sense to set platform toolset 140 along with sdk 8.1.

> where do you get the oldest SDK installed?

From the registry, see @JohannesMP comment.

So to answer all you other questions: basically I'm just saying that when action vs2017 is selected and the user is not specifically overriding the platform toolset and/or sdk version, Premake generates an invalid project file (invalid in the sense that it won't build out of the box) It does make sense in this case only (vs2017 action and no explicit platformtoolset/systemversion usage) to try to set the sdk version to 10 using @JohannesMP patch. Everyone would be happy : me because I no longer have to manually set the systemversion, and other users because it will still allow them to choose whatever combination of platform toolset and sdk they want.

Oh, and it would fix the problem of Premake not bootstrapping with vs2017 too :)

**Comment by tvandijck** [comment]
Maybe I'm missing something, but 
```
nmake -f Bootstrap.mak MSDEV=vs2017 windows-msbuild
```
Works perfectly fine for me for bootstrapping a Visual Studio 2017 build of Premake.

```
Done Building Project "D:\dev-premake\premake-core\build\bootstrap\Premake5.sln" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:46.35

D:\dev-premake\premake-core>
```

**Comment by tvandijck** [comment]
@starkos In my opinion the sensible default is to not emit the property.

If I create a new Console Application, it emits this to the vcxproj file.

```xml
<WindowsTargetPlatformVersion>10.0.16299.0</WindowsTargetPlatformVersion>
```

Which appears to be the latest I have installed on my desktop, and only available reliably through the registry, but it also makes the vcxproj file completely non-transferable. Anyone not having that SDK installed will have to update that property. Generating a Visual Studio project from linux/mac would also be impossible since no such registry setting is available.

I think to come up with a proper solution, more testing is required to see how Visual Studio deals with all these issues.

**Comment by dcourtois** [comment]
> I don't mind looking up the value where it is possible, but there needs to be a solution for people generating projects from Linux and macOS too.

But those people will generate a project that **does not compile** on Windows machines with Visual Studio 2017 and SDK 8.1 installed (because the generated project uses toolset v140 and default sdk, which is 8.1 when it's installed, and those 2 are incompatible, see my previous comments) ! 

Like @tvandijck says, the most sensible thing in my opinion is to try to look this up, **only if the user didn't explicitly set it, and only in the specific case of vs2017 with the v140 toolset** and if it's not found, then just do nothing at all : it's the current behavior, so you won't introduce any regression.

This proposition should only fix the basic cases where the projects are generated on the machine used to build, which I think is the majority of use cases. And for all the other cases, it will not change anything at all.

@tvandijck : I didn't know the MSDEV option to the bootstrap Makefile, I'll try that tomorrow, but I don't understand how it could correctly set the correct sdk version (since premake doesn't specify it, and the devenv /upgrade command used by the Makefile should not upgrade the sdk version since it's not there in the projects). Do you have both 8.1 and 10.0 SDKs installed ? Cause I think that if there is only the SDK 10 installed, Visual Studio will correctly set the correct platform version (this whole issue comes from having both SDKs installed, and Visual Studio 2017 using version 8.1 as its default, which is totally stupid, but ... :p)

**Comment by tvandijck** [comment]
@dcourtois yes I do have both 8.1 and 10 installed. I am however using the 'windows-msbuild' target, not the 'windows' target... the msbuild target skips the upgrade step, and uses msbuild directly to build, instead of msdev.. Although when specifying the MSDEV option, the upgrade should be a no-op.

**Comment by ratzlaff** [comment]
> but there needs to be a solution for people generating projects from Linux and macOS too.

I have nothing really to add except that we use this feature of premake at my company (just yesterday actually). It saves so much time when you are not a master of windows cmd.

**Comment by dcourtois** [comment]
@tvandijck Back to work on Windows, I just tried your command line, but unfortunately, it still doesn't compile. Here are the two lines I'm using:

```
D:\Development\Premake\premake-core> "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvars64.bat"
D:\Development\Premake\premake-core> nmake -f Bootstrap.mak MSDEV=vs2017 windows-msbuild
```

And the compilation fails with a ton of headers not found. I checked with the x86 version of vcvars, it's the same. And when I opened the solution with Visual Studio 2017 (`build\bootstrap\Premake5.sln`), all projects are incorrectly setup: Platform Toolset is v141, but Windows SDK Version is 8.1, leading to the broken build.

It's surprising that it builds fine for you ... would you mind telling me how toolset and sdk version are setup in your generated `build\bootstrap\*.vcxproj` projects ? And when you open the solution, what sdk version is selected ?

**Comment by samsinsane** [comment]
Out of curiosity, I looked into the issue of VS2017 generating a broken 8.1 install. Super simple fix - Install Windows 10.0.10240.0, both VS2015 and VS2017 use 10.0.10240.0 for some of the headers, specifically this folder: `C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt`. I noticed this on my tablet, which I thought had VS2017, but it doesn't - just VS2013 and VS2015.

Ultimately, both of these issues are bugs in Visual Studio, not Premake. Visual Studio installs half of the 8.1 SDK, Visual Studio picks the 8.1 SDK as the default even if it's not installed. Premake shouldn't start defaulting settings just because the tools contain bugs, especially when the default value changes depending on what is installed on the system.

Perhaps setting `systemversion "latest"` could do some magic and select the latest macOS, iOS and Windows versions, but users should be making the decision to do this.

Thinking about this a bit more, maybe Premake could provide a bit of a warning, more of a heads up, if it can't detect the 10.0.10240.0 and 8.1 SDKs and the project doesn't specify `systemversion`? Maybe if the project is being generated from a non-Windows system too and the `systemversion` isn't specified? Not super sold on that last bit, but an informative warning is more useful than attempting to build and getting told `stddef.h` doesn't exist.

**Comment by dcourtois** [comment]
> but users should be making the decision to do this

but the users currently have a broken build ... I utterly fail to see the point of forcing the user to update their scripts and specifically set a `systemversion "latest"` in a case where we know with absolute certainty that the solution won't build at all ! What's the point of not fixing something we know is broken ? Users already using `systemversion` will not be impacted, and the ones not using it will suddenly have something that works. And the ones that generate for vs2017 on another platform will still have a solution that doesn't build, but this is already the case.

No matter how I look at it, the worst case scenario is : it won't change a damn thing, the solution will still be broken, and in some cases : the solution will build out of the box. Where's the problem ?

Even CMake correctly sets the `WindowsTargetPlatformVersion` property in the generated .vcxproj files.

Note : I understand the concerns about starting to default some settings. But in this case the default are actually broken ! And I would also like to point out that Premake already does this for some other properties, like the toolset one for instance : `PlatformToolset` is actually set in the generated .vcxproj even if not setting is perfectly valid. So why defaulting that one, but not WindowsTargetPlatformVersion, especially when specifying the first and not the second result in a broken build ?

Note 2 : Leaving PlatformToolset empty in a .vcxproj falls back to v100 toolset. And this toolset is actually compatible with the default SDK Version (8.1) And using the retarget tool on such a project works perfectly. I guess this is why Premake's generated projects are broken : those 2 properties are not meant to be half specified : either set toolset and sdk version are do not set any of those.

**Comment by samsinsane** [comment]
@dcourtois I can build with both v141 and the 8.1 SDK. Just install the 10.0.10240.0 SDK and the 8.1 SDK will work. This is an issue with Visual Studio, not Premake.

**Comment by dcourtois** [comment]
@samsinsane What about Premake setting a default value for  `PlatformToolset` ? Why does it bother you so much to set a default value for `WindowsTargetPlatformVersion` when it doesn't for `PlatformToolset` ? I really don't understand :/

**Comment by tvandijck** [comment]
@dcourtois 

![image](https://user-images.githubusercontent.com/451515/32902928-8fe6e2fe-caa8-11e7-9b23-522edf959e9b.png)

As for what the right defaults are... I don't know, and am sort of indifferent to it, because for me it just works... But that is probably due to the fact that for work I have a need to install both VS2015 and VS2017, as well as all the appropriate SDKs (including Windows SDK 7.1a, because yes, we still support Windows XP for some of our games).

Running any of what I'm doing on a clean machine with just the latest VS2017 might yield very different results, I just don't know.

I'm completely fine with the registry fetch solution for the VS2017 action btw. It just won't work when running the action on linux/macOS.. which we just have to be OK with I guess.

As for the PlatformToolset, I also agree, not setting it actually sort of creates a broken project. It's particularly noticeable when creating Utility/Makefile projects.

**Comment by samsinsane** [comment]
> What about Premake setting a default value for `PlatformToolset` ?

That's a defined value for each version of Visual Studio. VS2010 is v100, VS2012 is v110, VS2013 is v120, VS2015 is v140 and VS2017 is v141. What's the value for `WindowsTargetPlatformVersion`? There isn't *one*, there's several and it will change again when they release the next version of the SDK. You're comparing something that's *fixed* to something that's *variable*.

> Why does it bother you so much to set a default value for `WindowsTargetPlatformVersion` when it doesn't for `PlatformToolset` ? I really don't understand :/

Reproducible builds. When I run `premake5.exe vs2017`, I expect to get the exact same project file every single time. We've had several patches from a few contributors to ensure this occurs. If I update VS2017 and that happens to bring in a new version of the Windows SDK, running `premake5.exe vs2017` *shouldn't* generate a different project file. I didn't change anything in my Premake scripts, I didn't change the Premake binary, nor did I make the conscious decision to have Premake rely on my development environment setup.

Visual Studio bugs shouldn't be fixed by Premake. VS2017 Installer didn't install all of the 8.1 SDK. VS2017 doesn't appear to have a default WinSDK that's guaranteed to be installed. If it did, I would 100% agree that we should be setting it, but to my knowledge it doesn't have one. I would love to be proven wrong about this though.

**Comment by tvandijck** [comment]
@samsinsane You know, I sort of agree, but also am not really bothered by the fact that it might not generate 100% the same thing all the time...

The workflow or requirements that Jason aims for, where you can generate a project/solution somewhere, zip it up, and then transfer it across the world to someone without premake that then can build your project, is a amicable goal, but a workflow I personally see no value in.

For us here at Blizzard, and for me personally at home, we don't commit vcxproj/makefiles to our repositories, our build farms run Premake, and then build that generated solution. Our package system has a global cache setup on those build farms, and because of that shared space, most of our projects actually contain absolute paths anyway, so the transfer-ability of the projects is entirely lost already anyway. I don't have that desire basically..

I want premake to be deterministic yes, but for my current environment, and if that environment changes I somewhat expect premake to adjust as needed.

One of our biggest problems right now actually is that the generated makefiles sometimes are simply incorrect... If I run on Centos5, I get gcc44, which doesn't support a couple of compiler options premake generated, because those were added in gcc48... While I totally understand that Centos5 is 20 years old, and as a community we shouldn't have to care about that anymore, but it's a problem for us here.

I wish premake would simply run 'gcc --version', or 'clang --version', and adjusted its output based on what it finds out.. And I guess that wish is somewhat similar to what can be discovered for installed SDK's on Windows.

I think ultimately this comes down to how you use Premake, and how we as a community see value in the different use-cases. **Transfer-ability** being the key feature we're basically arguing over here.

**Comment by samsinsane** [comment]
> For us here at Blizzard, and for me personally at home, we don't commit vcxproj/makefiles to our repositories, our build farms run Premake, and then build that generated solution.

Same. The only time I've used Premake to generate for a different platform is when I'm at my workstation, trying to figure out what the hell is going on with one of the other builds.

> I want premake to be deterministic yes, but for my current environment, and if that environment changes I somewhat expect premake to adjust as needed.

I somewhat agree. However, these two things aren't exactly equal. Your environment changing and installing an **additional** Windows SDK aren't the same thing. I elaborate on this below, I didn't intend to because it's near 4am, but I did, oops.

> One of our biggest problems right now actually is that the generated makefiles sometimes are simply incorrect... If I run on Centos5, I get gcc44, which doesn't support a couple of compiler options premake generated, because those were added in gcc48... While I totally understand that Centos5 is 20 years old, and as a community we shouldn't have to care about that anymore, but it's a problem for us here.

I had migrated our SDK from an amalgamation of committed VS projects/solutions, and a JamPlus "configuration file" to Premake months ago (it could have been last year, I've lost track of work time). I had to wrangle Premake to emit VS2010->VS2015 for Windows, VS2013 (or maybe VS2015) for WinRT, XCode or Makefiles for macOS and iOS (I don't remember what I did here at all), Makefiles for Ubuntu 14.04 and 16.04, CentOS 6, and 7 *and* Android. Thanks to CentOS, I know the pitfalls of `toolset "gcc"`, so I kind of feel your pain. Thankfully, we managed to phase out Ubuntu 12.04 and CentOS 5 while I was migrating this project from Perforce to Git months before the Premake migration mostly because I didn't have a Ubuntu 12.04 or CentOS5 Docker image, and we didn't have the IT resources for all the other stuff that comes with supporting older versions. I'm sorry you still have to deal with CentOS5. 😢 

> I wish premake would simply run 'gcc --version', or 'clang --version', and adjusted its output based on what it finds out.. And I guess that wish is somewhat similar to what can be discovered for installed SDK's on Windows.

I see some minor issues with this. Your version specific adjustments make sense, sometimes flags aren't available, but your C++11 code will generally always compile regardless of certain flags right? Maybe an optimization level is missing in an earlier GCC version, but would the build fail because you used a different one? No. Would the build fail because you used a different version of an SDK? Depends. Maybe the build won't fail, maybe it just won't operate like it did on your machine. Or maybe the function won't exist anymore, or it's just a stub returning an error code. Speed/optimizations vs functionality are two different things, and I feel like that's kind of being lost here. There's a huge difference between using `-O1` and `-O2`, or whatever the optimization flags are, and using `-std=c++11` and `-std=c++17`. The latter is more inline with SDKs. 

> **Transfer-ability** being the key feature we're basically arguing over here.

Well, that and the ability to just run todays code in a couple of years with the same things installed and not have it pick up the "VS16" Windows 10 SDK that mightn't be compatible with VS2017 for whatever reason.

**Comment by ratzlaff** [comment]
The best 100% most awesome feature of premake is that it does not require a working environment to do its job. If the project it makes for me does not work, then I need to either:
1) fix my environment
or
2) add more specific configuration {...} areas

Having premake be deterministic is **awesome** and (in my opinion) not something that should be trivially lost.

**Comment by tvandijck** [comment]
@ratzlaff Again I somewhat agree... but neither option 1 or 2 are sometimes viable...

For example "-Og" is an unsupported compiler flag in gcc44, it simply makes the build fail with an error "unknown argument".

Premake however doesn't have any kind of specific configuration for that.... you can't do
```lua
filter { 'toolset:gcc44' }
```

because the toolset is simply 'gcc', without any kind of version specified.
neither does premake even know what version of gcc is being targeted.
And -Og is generated by the `optimize` API anyway, which is hardcoded in the gmake action, and hence totally out of my control.

Essentially, I can't even do anything about it at the project level, and so other then fixing the code there is no option 1 or 2.
https://github.com/Blizzard/premake-core/commit/44cced5cb859b6078af9e3823b88162fcda2ffeb

**Comment by dcourtois** [comment]
> Reproducible builds. When I run premake5.exe vs2017, I expect to get the exact same project file every single time.

Ok, it does make sense, thanks for elaborating :)

Then @samsinsane's suggestion (`systemversion "latest"` or something along this line, which would try to fetch the latest SDK 10 version from the registry, when vs2017 is used) would be a pretty good compromise : users in a simple environment will be able to get a working project with minimum modifications of their scripts, and Premake will remain deterministic.

Would that be acceptable for everyone ?

**Comment by starkos** [comment]
I'm fine with `systemversion "latest"`. I'm also okay with looking up the value if it is available.

So…what does Visual Studio 2017 do on its own? Taking Premake out of this for a moment…there are going to be lots of people creating projects, checking them into GitHub, and sharing them with other people. What happens @dcourtois tries to load a project created by @tvandijck? What if it tries to load a project made six months and a couple of platform updates ago? Does it really just fall over?

I still feel like there ought to some kind of sensible default to put there which will at least allow VS to do the least bad thing.

**Comment by dcourtois** [comment]
1. if your create a new C++ project inside Visual Studio, it will set the latest installed SDK10 version and toolsets, which is on my machine:

```
<WindowsTargetPlatformVersion>10.0.16299.0</WindowsTargetPlatformVersion>
<PlatformToolset>v141</PlatformToolset>
```

Note that the version is the one found in the registry `Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Microsoft SDKs\Windows\v10.0`'s `ProductVersion` key.

2. Resetting (via the project properties inside vs, or by removing the entry in the .vcxproj directly) `Windows SDK Version` and `Platform Toolset` respectively chooses the SDK 8.1 and toolset v100, no matter the value of the other. Both settings are completely independent from one another (which leads to the discussed problem when setting toolset and not sdk version)

3. If I set the `WindowsTargetPlatformVersion` property to a valid version number which is not installed on my machine, the build fails. Basically it means that if I open a project generated with either an uninstalled SDK version or no version at all, it will fail to build. What this means if @tvandijck were to generate a file, and I was building it :
    1. `systemversion` was not used, SDK 8.1 is chosen, build fails (unless I have the exact version of the SDK 10 mentioned by @samsinsane , but I for one do not have it, and since it's not the latest one it won't be installed with recent installs of VS2017)
    2. `systemversion "some_version"` is used : if I have the specific version installed, fine, if I don't, build fails.
    3. `systemversion "latest"` is used : same as previously.

To sum up, it seems there is absolutely no way to tell Visual Studio "Hey I want to use the latest installed SDK 10, choose the correct version for me please !" unfortunately (at least I haven't found one)

**Comment by starkos** [comment]
Continuing to leave Premake out of it, and just working with Visual Studio only, my understanding of the problem is: 

- I open Visual Studio, and File > New Project…  
- Visual Studio writes "10.0.16299.0" into the project
- I check all of this into git and forget about it for a while
- Time goes by, platform version 10.1.34893.1 is released
- I blow the dust off my old project and fire it up
- Visual Studio borks and falls over

Have I got that right? If so, what's the resolution (again, leaving Premake out of it)? Does the project load at all? Does it just fail to build? How do you fix the project so it works correctly again?

If I've understood this correctly, it going to be just a matter of time before a fix is provided and we'll just have to ride it out. From what I've been able to infer, my current opinion on the VS'17 exporter is:

- If no specific `systemversion` is set, or if it is set to "latest", and we're running on Windows, try to read the registry key. If we get a valid value, use it.

- If we don't have a key, put in reasonable default values. It won't build, but it's all we've got.

- Consider showing a warning message when this happens, so no one gets surprised. 

For Makefiles I'm not so sure. Can we (ick) shell out to `reg query` to get the value at build time?

**Comment by tvandijck** [comment]
If the project contains an version you have no installed you get:

```
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\IDE\VC\VCTargets\Platforms\x64\PlatformToolsets\v141\Toolset.targets(36,5): error MSB8036: The Windows SDK version 10.0.16298.0 was not found. Install the required version of Windows SDK or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solution".
1>Done building project "ConsoleApplication12.vcxproj" -- FAILED.
```

You can then fix the project by clicking retarget.
![image](https://user-images.githubusercontent.com/451515/32959152-89c4b086-cb75-11e7-93cc-34d374b93dcf.png)

![image](https://user-images.githubusercontent.com/451515/32959166-93ad02f6-cb75-11e7-9fb7-c1432e8e77df.png)

In my example here there is only one vcxproj, but you can effectively retarget the entire solution in a single click.


So all in all, basically looking up the value in the registry and setting it to latest is pretty much what visual studio itself does as well at creation time. Premake should just do the same, unless the value is specified explicitly.... `systemversion 'latest'` seems like a reasonable solution, but I would also just make that the default behavior, and I would go so far as to say that I'd like this to apply to both VS2015 and VS2017, since both expose the property.

**Comment by dcourtois** [comment]
Hi there ! Should I try implementing the `systemversion "latest"` and submit it as a pull request, if everyone agrees on this syntax ?

**Comment by tvandijck** [comment]
yeah, that would be appreciated...

**Comment by samsinsane** [comment]
Closing this as #994 added `systemversion "latest"` and has been merged in.
